src/inmanta/parser/plyInmantaParser.py

```diff
@@ -799,20 +806,36 @@ def p_constants_collect(p: YaccProduction) -> None:
     p[0] = p[3]
 
 
-def p_param_list_collect(p: YaccProduction) -> None:
-    """param_list : ID '=' operand ',' param_list"""
-    p[5].insert(0, (p[1], p[3]))
-    p[0] = p[5]
+def p_wrapped_kwargs(p: YaccProduction) -> None:
+    "wrapped_kwargs : '*' '*' operand"
+    p[0] = WrappedKwargs(p[3])
 
 
-def p_param_list_term(p: YaccProduction) -> None:
-    "param_list : ID '=' operand"
-    p[0] = [(p[1], p[3])]
+def p_param_list_element_explicit(p: YaccProduction) -> None:
+    "param_list_element : ID '=' operand"
+    p[0] = ((p[1], p[3]), None)
 
 
-def p_param_list_term_2(p: YaccProduction) -> None:
-    "param_list : "
-    p[0] = []
+def p_param_list_element_kwargs(p: YaccProduction) -> None:
+    "param_list_element : wrapped_kwargs"
+    p[0] = (None, p[1])
+
+
+def p_param_list_empty(p: YaccProduction) -> None:
+    """param_list : param_list_empty
+        param_list_empty : empty"""
+    p[0] = ([], [])
+
+
+def p_param_list_nonempty(p: YaccProduction) -> None:
+    """param_list : param_list_element empty param_list_empty
```

Comment:
why ```empty param_list_empty``` ? 

(why is that even possible?)

Reply:
It works because empty matches the empty string. The upside of this approach is that it provides abstraction: there's only one parser rule needed for the nonempty param_list instead of one for a single-element list and another one for a multiple-element list, both with essentially the same body.
Besides that, I'm not sure if this is good practice. This is one of the things I was referring to when I asked for a critical review. My opinion is that the abstraction it provides more than balances the slightly unusual formulation of the rule. What are your thoughts on this?

Reply:
perhaps add a comment